src/OpenCloudDriver.php

```diff
@@ -0,0 +1,267 @@
+{
+    /**
+     * @var QueueService
+     */
+    private $service;
+    /**
+     * Default TTL for published messages / claims
+     * @var int
+     */
+    private $ttl;
+    /**
+     * @var Queue[]
+     */
+    private $queues = [];
+    /**
+     * The claims we are currently working on
+     * @var Message[]
+     */
+    private $claims = [];
+    /**
+     * @var \OpenCloud\Common\Collection\PaginatedIterator[]
+     */
+    private $lists = [];
+
+    /**
+     * Constructor
+     *
+     * @param QueueService $service
+     * @param int $prefetch
+     * @param int $ttl
+     */
+    public function __construct(QueueService $service, $queues = [], $prefetch = null, $ttl = Claim::TTL_DEFAULT)
+    {
+        parent::__construct($prefetch);
+
+        $this->service = $service;
+        $this->ttl = $ttl;
+
+        $this->populateQueues($queues);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function listQueues()
+    {
+        $result = [];
+        /* @var $queues Queue[] */
+        if ($queues = $this->service->listQueues()) {
+            foreach ($queues as $queue) {
+                $result[] = $queue->getName();
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function createQueue($queueName)
+    {
+        $this->service->createQueue($queueName);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function countMessages($queueName)
+    {
+        try {
+            $queue = $this->getQueue($queueName, false);
+            if ($stats = $queue->getStats()) {
+                return $stats->total;
+            }
+        } catch (BadResponseException $e) {
+            if ($e->getResponse()->getStatusCode() != 404) {
+                throw $e;
+            }
+        }
+        return 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function pushMessage($queueName, $message)
+    {
+        $queue = $this->getQueue($queueName);
+        $queue->createMessage([
+            'ttl' => $this->ttl,
+            'body' => $message
+        ]);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function popMessage($queueName, $interval = 5)
+    {
+        /* @var Message $claim */
+        if ($message = $this->cache->pop($queueName)) {
+            return $message;
+        }
+
+        $queue = $this->getQueue($queueName);
+
+        $runtime = microtime(true) + $interval;
+        while (microtime(true) < $runtime) {
+            if ($claims = $this->claimMessages($queue)) {
+                foreach ($claims as $claim) {
+                    $this->claims[$claim->getHref()] = $claim;
+                    $this->cache->push($queueName, [$claim->getBody(), $claim->getHref()]);
+                }
+
+                if (count($claims)) {
+                    return $this->cache->pop($queueName);
+                }
+            }
+
+            //sleep for 10 ms
+            usleep(1000);
```

Comment:
OK, but wouldn't that mean it'll only ever loop once? The Doctrine, Mongo and Amqp drivers all `usleep(10000)`.